skbio/stats/distance/_pairwise_distances.py

```diff
@@ -0,0 +1,446 @@
+                            % type(distances).__name__)
+
+        distances_ = {}
+        ids = set()
+        for id_pair, distance in distances.items():
+            if not is_id_pair(id_pair):
+                raise TypeError(
+                    "Each key in `distances` must be a tuple of exactly two "
+                    "strings, not %r" % (id_pair,))
+
+            if not isinstance(distance, float):
+                raise TypeError(
+                    "Each distance in `distances` must be a float, not type %r"
+                    % type(distance).__name__)
+
+            if np.isnan(distance):
+                raise ValueError("`distances` contains a distance that is NaN")
+
+            id1, id2 = id_pair
+            if id1 == id2:
+                if distance != 0.0:
+                    raise ValueError(
+                        "`distances` contains a nonzero distance between ID "
+                        "%r and itself: %r" % (id1, distance))
+            elif (id2, id1) in distances_:
+                if distance != distances_[(id2, id1)]:
+                    raise ValueError(
+                        "`distances` contains an asymmetric distance between "
+                        "IDs %r and %r: %r != %r" % (id1, id2, distance,
+                                                     distances_[(id2, id1)]))
+            else:
+                distances_[id_pair] = distance
+
+            ids.add(id1)
+            ids.add(id2)
+
+        self._distances = distances_
+        self._ids = frozenset(ids)
+
+    @property
+    @experimental(as_of="0.4.2-dev")
+    def ids(self):
+        """Set of IDs present in this ``PairwiseDistances``.
+
+        Returns
+        -------
+        frozenset
+            Set of IDs present in this ``PairwiseDistances``.
+
+        Notes
+        -----
+        This property is not writeable.
+
+        Examples
+        --------
+        >>> from skbio.stats.distance import PairwiseDistances
+        >>> pwdist = PairwiseDistances({('A', 'B'): 0.1, ('C', 'C'): 0.0})
+        >>> pwdist.ids == frozenset({'A', 'B', 'C'})
+        True
+
+        """
+        return self._ids
+
+    @experimental(as_of="0.4.2-dev")
+    def __bool__(self):
+        """Boolean indicating whether this ``PairwiseDistances`` is empty.
+
+        Returns
+        -------
+        bool
+            ``False`` if there are no "within" or "between" distances, ``True``
+            otherwise.
+
+        Examples
+        --------
+        >>> from skbio.stats.distance import PairwiseDistances
+        >>> pwdist = PairwiseDistances({('A', 'B'): 0.1, ('C', 'C'): 0.0})
+        >>> bool(pwdist)
+        True
+        >>> pwdist = PairwiseDistances({})
+        >>> bool(pwdist)
+        False
+
+        """
+        return len(self) > 0
+
+    @experimental(as_of="0.4.2-dev")
+    def __len__(self):
+        """Number of "within" and "between" pairwise distances.
```

Comment:
A philosophical question: why do we care about the "within" distances at all in this object? It seems like a lot of work for some relatively counter-intuitive behaviour, especially this method and `__iter__`.

I would kind of expect a partial strategy to skip the diagonal anyhow. Wouldn't it make just as much sense for `DistanceMatrix` to complain if it actually found a "within" distance that wasn't 0, and otherwise it hallucinates them in the alternate constructer?